### PR TITLE
Factor tableIndex out of tableReader

### LINKED ISSUE #2949: (NBS) TablePersister leaks s3 table data
https://github.com/attic-labs/noms/blob/master/go/nbs/table_persister.go#L73 is attempting to avoid re-reading the index, but ends up keeping the entire table alive.

### PR DESCRIPTION
**Pull request by rafael-atticlabs**
Fixes #2949. Towards #2948.

### CODE REVIEW

**Comment by rafael-atticlabs** [comment]
The next step here will be to have a cache of `tableIndex` stored on AWSStoreFactory.

File: go/nbs/table_reader.go [hunk 1]
```diff
@@ -80,6 +89,55 @@ func computePrefixes(count uint32, buff []byte) (prefixes []uint64, ordinals []u
 	return
 }
 
+func (ti tableIndex) prefixIdxToOrdinal(idx uint32) uint32 {
```
Comment (rafael-atticlabs) [inline go/nbs/table_reader.go]: the only code that changed here is `lookupOrdinal`

File: go/nbs/table_reader.go [hunk 2]
```diff
@@ -14,20 +14,22 @@ import (
...
-func newTableReader(buff []byte, r io.ReaderAt, readAmpThresh uint64) tableReader {
-	tr := tableReader{r: r, readAmpThresh: readAmpThresh}
+// tableReader implements get & has queries against a single nbs table. goroutine safe.
+type tableReader struct {
+	tableIndex
+	r             io.ReaderAt
+	readAmpThresh uint64
+}
 
+// parses a valid nbs table index byte stream and returns a reader. |buff| must end with an NBS index and footer, though it may contain an unspecified number of bytes before that data. |tableIndex| doesn't keep alive any references to |buff|.
```
Comment (cmasone-attic) [inline go/nbs/table_reader.go]: This returns a tableIndex, not a reader

File: go/nbs/table_reader.go [hunk 3]
```diff
@@ -38,21 +40,28 @@ func newTableReader(buff []byte, r io.ReaderAt, readAmpThresh uint64) tableReade
...
-	tr.chunkCount = binary.BigEndian.Uint32(buff[pos:])
+	chunkCount := binary.BigEndian.Uint32(buff[pos:])
 
 	// index
-	suffixesSize := uint64(tr.chunkCount) * addrSuffixSize
+	suffixesSize := uint64(chunkCount) * addrSuffixSize
 	pos -= suffixesSize
-	tr.suffixes = buff[pos : pos+suffixesSize]
+	suffixes := make([]byte, suffixesSize)
+	copy(suffixes, buff[pos:])
```
Comment (cmasone-attic) [inline go/nbs/table_reader.go]: Per my comment in slack, "I do also wonder if the fact that different approaches are ideal for different persistence layers (e.g. mmap the index for files, copying for S3) means that we should have different code pathways for creating the tableReader"

There's no need to copy this data over when backed by files, though there may be a need when the data is already in memory.

File: go/nbs/table_reader.go [hunk 4]
```diff
@@ -80,6 +89,55 @@ func computePrefixes(count uint32, buff []byte) (prefixes []uint64, ordinals []u
...
+	li := uint64(ordinal) * addrSuffixSize
+	return bytes.Compare(h[addrPrefixSize:], ti.suffixes[li:li+addrSuffixSize]) == 0
+}
+
+// returns the ordinal of |h| if present. returns |ti.chunkCount| if absent
+func (ti tableIndex) lookupOrdinal(h addr) uint32 {
+	prefix := h.Prefix()
+	idx := ti.prefixIdx(prefix)
+
+	for ; idx < ti.chunkCount && ti.prefixes[idx] == prefix; idx++ {
```
Comment (cmasone-attic) [inline go/nbs/table_reader.go]: Why can't `idx` be initialized inline in the for loop?

File: go/nbs/table_reader.go (hunk 2 above)
Comment (rafael-atticlabs) [inline go/nbs/table_reader.go]: done

File: go/nbs/table_reader.go (hunk 3 above)
Comment (rafael-atticlabs) [inline go/nbs/table_reader.go]: yeah, i'm hoping we'll get to the point that a table index is just a wrapper around `[]byte` and at that point, it's much easier to have different approaches to managing the lifetime of those slices. but for now, i'd rather just have one approach which is that the index always makes a copy and keeps it in memory.

File: go/nbs/table_reader.go (hunk 4 above)
Comment (rafael-atticlabs) [inline go/nbs/table_reader.go]: done

**Comment by rafael-atticlabs** [comment]
ptal

File: go/nbs/table_reader.go (hunk 3 above)
Comment (cmasone-attic) [inline go/nbs/table_reader.go]: Per offline conversation, it seems that the goal is to make the table index a wrapper around a memory-mapped piece of a file no matter what the ultimate storage backend is. That sounds fine to me.

**Pull request merged**